tokio-util/src/codec/framed_impl.rs

```diff
@@ -120,42 +120,101 @@ where
+        // In fact, most state transitions occur with a return.
+        // The intitial state is `reading`.
+        //
+        // | state   | eof   | is_readable |
+        // |---------|-------|-------------|
+        // | reading | false | false       |
+        // | framing | false | true        |
+        // | pausing | true  | true        |
+        // | paused  | true  | false       |
+        //
+        //                                 `decode_eof`
+        //                                 returns `Some`             read 0 bytes
+        //                                   │       │                 │      │
+        //                                   │       ▼                 │      ▼
+        //                                   ┌───────┐ `decode_eof`    ┌──────┐
+        //                 ┌──read 0 bytes──▶│pausing│─returns `None`─▶│paused│──┐
+        //                 │                 └───────┘                 └──────┘  │
+        // pending read┐   │                     ┌──────┐                  │  ▲  │
+        //      │      │   │                     │      │                  │  │  │
+        //      │      ▼   │                     │  `decode` returns `Some`│ pending read
+        //      │  ╔═══════╗                 ┌───────┐◀─┘                  │
+        //      └──║reading║─read n>0 bytes─▶│framing│                     │
+        //         ╚═══════╝                 └───────┘◀──────read n>0 bytes┘
+        //             ▲                         │
+        //             │                         │
+        //             └─`decode` returns `None`─┘
         loop {
-            // Repeatedly call `decode` or `decode_eof` as long as it is
-            // "readable". Readable is defined as not having returned `None`. If
-            // the upstream has returned EOF, and the decoder is no longer
-            // readable, it can be assumed that the decoder will never become
-            // readable again, at which point the stream is terminated.
+            // Repeatedly call `decode` or `decode_eof` as long as long as the buffer
+            // is "readable". Readable means that it _might_ contain data that could be consumed
+            // by `decode` or `decode_eof`. Both `decode` and `decode_eof` signal that the buffer
+            // is no longer readable for them by returning `None`.
+            // If `decode` signals that it couldn't read from the buffer and the upstream has
+            // returned eof, `decode_eof` will attemp to decode the remainder as closing frames.
+            // The buffer might become readable again if the underlying AsyncRead is a FIFO or file.
+            // However, we still want to make sure that upon encountering an eof
+            // we actually finish emmiting all it's associated `decode_eof` frames.
+            // Furthermore, we don't want to emit any `decode_eof` frames on retried
+            // reads after an eof unless we've actually read more data.
```

Comment:
Can you reflow this text so the line length is more consistent? And when you want an explicit line break between paragraphs, please put an empty line between them.

Reply:
gonna get to it asap 👍

Reply:
I've cleaned up and shortened the docs 😄